controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -18,30 +20,47 @@
 import io.pravega.controller.store.stream.tables.Data;
 import io.pravega.controller.store.stream.tables.HistoryRecord;
 import io.pravega.controller.store.stream.tables.IndexRecord;
-import io.pravega.controller.store.stream.tables.SegmentRecord;
 import io.pravega.controller.store.stream.tables.State;
 import io.pravega.controller.store.stream.tables.TableHelper;
-import io.pravega.client.stream.StreamConfiguration;
 import lombok.extern.slf4j.Slf4j;
 import lombok.val;
+import org.apache.commons.lang.SerializationUtils;
 import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
 
 import java.util.AbstractMap;
 import java.util.AbstractMap.SimpleImmutableEntry;
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 @Slf4j
 public abstract class PersistentStreamBase<T> implements Stream {
 
+    private static final Map<State, List<State>> STATE_TRANSITIONS;
+
+    static {
+        Map<State, List<State>> map = new HashMap<>();
```

Comment:
The Enum is OK either way, but the race does not require modifying the map. Merely that there is one thread that modified the map some time in the past and no memory barrier has been encountered. If this is the case even a get() is not safe because it may not see the update applied or worse it could see it partially applied and bad things could (as in the example above where it caused the caller to go into an infinite loop) See: https://en.wikipedia.org/wiki/Java_memory_model